### PR TITLE
Hoist class-private vars above the class; move static class calls to …

### PR DESCRIPTION
**Pull request by GoodForOneFare**
…below the class.

This works for Shopify's codebase because we have one class per file, and don't do anything weird with private-static calls.  If it'd help to get this PR accepted, I can add a command line flag to enable/disable this behaviour.

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -3,7 +3,7 @@
   "env": {
     "browser": true,
     "jest": true,
-    "jasmine": true,
+    "mocha": true,
```
Comment (juliankrispel) [inline .eslintrc]: why this change?

Comment (GoodForOneFare) [inline .eslintrc]: `mocha` allows `before`/`after` without flagging them as undefined.  I'm using those to suppress warnings about private vars/calls in the test output.

**Comment by juliankrispel** [comment]
Hm. @GoodForOneFare I think a flag would be nice.

Also, how does this deal with naming collisions, I'd like to make sure that these variable names don't collide with any other variable names, since they're mounted in the same scope as the class definition.

**Comment by GoodForOneFare** [comment]
At the moment it only warns about naming collisions.  I can take a look at throwing exceptions.

Any preference for a flag name?  `transform_class_private_statements=[true|false]` with a default of `false`?

**Comment by juliankrispel** [comment]
@GoodForOneFare `decaf --private-class-statements` with a boolean :+1:

thanks dude

**Comment by juliankrispel** [comment]
@GoodForOneFare no need to throw exceptions, instead you could use the coffeescript helper attached to the meta object, `meta.scope.freeVariable` >> https://github.com/rainforestapp/decaf/blob/master/src/parser.js#L202

**Comment by ukd1** [comment]
@GoodForOneFare @juliankrispel what's the status on this?

**Comment by GoodForOneFare** [comment]
I have a work in progress that's most of the way there, but other stuff came up.  I'll take a look this weekend to pare down what I have and get a PR up.

**Comment by juliankrispel** [comment]
@GoodForOneFare I'd be up for pairing if you wanna

**Comment by ukd1** [comment]
@GoodForOneFare @juliankrispel is this gonna happen?

**Comment by ukd1** [comment]
I'm gonna close this. Please re-open @juliankrispel if you get time
